autonat/autonat-v2.md

```diff
@@ -0,0 +1,193 @@
+resource limit reached or blacklisting.
+
+`E_TRANSPORT_NOT_SUPPORTED`: the peer didn't have the ability to dial any of the
+requested addresses.
+
+`E_BAD_REQUEST`: the peer didn't attempt a dial because it was unable to decode
+the message. This includes inability to decode the requested address to dial.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+
+### Consideration for DDOS Prevention
+
+In order to prevent attacks like the one described in [RFC 3489, Section
+12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
+below), implementations MUST NOT dial any multiaddress unless it is based on the
+IP address the requesting node is observed as. This restriction as well implies
+that implementations MUST NOT accept dial requests via relayed connections as
+one can not validate the IP address of the requesting node.
+
+> RFC 3489 12.1.1 Attack I: DDOS Against a Target
+>
+> In this case, the attacker provides a large number of clients with the same
+> faked MAPPED-ADDRESS that points to the intended target. This will trick all
+> the STUN clients into thinking that their addresses are equal to that of the
+> target. The clients then hand out that address in order to receive traffic on
+> it (for example, in SIP or H.323 messages). However, all of that traffic
+> becomes focused at the intended target. The attack can provide substantial
+> amplification, especially when used with clients that are using STUN to enable
+> multimedia applications.
+
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+
+## RPC Messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec]. 
+
+All RPC messages are of type `Message`. A `DialRequest` message is sent as a
+`Message` with the `dialRequest` field set and the `type` field set to
+`DIAL_REQUEST`. Other message types `DialResponse` and `DialAttempt` are handled
+similarly.
+
+
+```proto
+syntax = "proto3";
+
+message Message {
+  enum MessageType {
+    DIAL_REQUEST  = 0;
+    DIAL_RESPONSE = 1;
+    DIAL_ATTEMPT  = 2;
+  }
+
+  MessageType type          = 1;
+  DialRequest dialRequest   = 2;
+  DialResponse dialResponse = 3;
+  DialAttempt dialAttempt   = 4;
+}
+
+message AddressDialRequest {
+  bytes addr = 1;
+  fixed64 nonce = 2;
+}
+
+message DialRequest {
+  repeated AddressDialRequest addressDialRequests = 1;
+}
```

Comment:
A different naming idea:

```suggestion
message Candidates {
  bytes addr = 1;
  fixed64 nonce = 2;
}

message DialRequest {
  repeated Candidates candidates = 1;
}
```

I've found the repetition of the words "addresses" and "request" hard for creating a mental model of what is happening upon first read.

Reply:
I like this suggestion! thanks!